internal/models/config.go

```diff
@@ -0,0 +1,171 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package models defines the data models for entities. This file defines the
+// models for V2 configuration format. The defined entities are:
+// ConfigV2 (the top-level structure), Receivers, Exporters, Options, Pipelines.
+package models
+
+/*
+Receivers, Exporters and Options typically have common configuration settings, however
+sometimes specific implementations (e.g. StackDriver) will have extra configuration settings.
+This requires the configuration data for these entities to be polymorphic.
+
+To satisfy these requirements we declare interfaces ReceiverCfg, ExporterCfg, OptionCfg,
+which define the behavior. We also provide helper structs ReceiverCommon, ExporterCommon,
+OptionCommon, which define the common settings and un-marshaling from config files.
+
+Specific Receivers/Exporters/Options are expected to at the minimum implement the
+corresponding interface and if they have additional settings they must also extend
+the corresponding common settings struct (the easiest approach is to embed the common struct).
+
+For example usage see tests.
+*/
+
+// ConfigV2 defines the configuration V2 for the various elements of collector or agent.
+type ConfigV2 struct {
+	Receivers []ReceiverCfg `mapstructure:"-"`
+	Exporters []ExporterCfg `mapstructure:"-"`
+	Options   []OptionCfg   `mapstructure:"-"`
+	Pipelines Pipelines     `mapstructure:"pipelines"`
+}
+
+// ReceiversKeyName is the configuration key name for receivers section.
+const ReceiversKeyName = "receivers"
+
+// ExportersKeyName is the configuration key name for exporters section.
+const ExportersKeyName = "exporters"
+
+// OptionsKeyName is the configuration key name for options section.
+const OptionsKeyName = "options"
+
+// Pipelines defines the pipelines section of configuration.
+type Pipelines struct {
+	Traces  []*Pipeline `mapstructure:"traces"`
+	Metrics []*Pipeline `mapstructure:"metrics"`
+}
+
+// Pipeline defines a single pipeline.
+type Pipeline struct {
+	Name       string              `mapstructure:"name"`
+	Receivers  []PipelineReceiver  `mapstructure:"receivers"`
+	Operations []PipelineOperation `mapstructure:"ordered-operations"`
+	Exporters  []PipelineExporter  `mapstructure:"exporters"`
+}
+
+// PipelineReceiver defines a receiver of a pipeline.
+type PipelineReceiver struct {
+	Name string `mapstructure:"name"`
+}
+
+// PipelineOperation defines an operation of a pipeline.
+type PipelineOperation struct {
+	Name string `mapstructure:"name"`
+}
+
+// PipelineExporter defines an exporter of a pipeline.
+type PipelineExporter struct {
+	Name string `mapstructure:"name"`
+}
+
+// NamedEntity defines a configuration entity that has a name.
+type NamedEntity interface {
```

Comment:
Why interface and not just?
```go
type NamedEntity struct {
  Name string `mapstructure:"name"`
}
```

Reply:
This is an interface that is embedded in other interfaces (e.g. `ReceiverCfg`). If I make it a `struct` I cannot embed it in an interface.

Reply:
Why not having all Structs then?

Reply:
I need `ReceiverCfg`,`ExporterCfg` and `ProcessorCfg` to be polymorphic (because e.g. different receivers have different configuration settings). I need them to be interfaces, structs will not work.